disperser/apiserver/server.go

```diff
@@ -257,45 +273,73 @@ func (s *DispersalServer) disperseBlob(ctx context.Context, blob *core.Blob) (*p
 	}, nil
 }
 
-func (s *DispersalServer) getAccountRate(origin string, quorumID core.QuorumID) (*PerUserRateInfo, error) {
+type RateKeys struct {
+	ThroughputKey string
+	BlobRateKey   string
+}
+
+func (s *DispersalServer) getAccountRate(origin, authenticatedAddress string, quorumID core.QuorumID) (*PerUserRateInfo, *RateKeys, error) {
 	unauthRates, ok := s.rateConfig.QuorumRateInfos[quorumID]
 	if !ok {
-		return nil, fmt.Errorf("no configured rate exists for quorum %d", quorumID)
+		return nil, nil, fmt.Errorf("no configured rate exists for quorum %d", quorumID)
+	}
+
+	// Check if the origin is in the allowlist
+	rates := &PerUserRateInfo{
+		Throughput: unauthRates.PerUserUnauthThroughput,
+		BlobRate:   unauthRates.PerUserUnauthBlobRate,
+	}
+
+	keys := &RateKeys{
+		ThroughputKey: "ip:" + origin,
+		BlobRateKey:   "ip:" + origin,
 	}
 
-	for ip, rateInfoByQuorum := range s.rateConfig.Allowlist {
-		if !strings.Contains(origin, ip) {
+	for account, rateInfoByQuorum := range s.rateConfig.Allowlist {
+		if !strings.Contains(origin, account) {
 			continue
 		}
 
 		rateInfo, ok := rateInfoByQuorum[quorumID]
 		if !ok {
-			continue
+			break
 		}
 
-		throughput := unauthRates.PerUserUnauthThroughput
-		if rateInfo.Throughput > 0 {
-			throughput = rateInfo.Throughput
+		if rateInfo.Throughput > unauthRates.PerUserUnauthThroughput {
+			rates.Throughput = rateInfo.Throughput
 		}
 
-		blobRate := unauthRates.PerUserUnauthBlobRate
-		if rateInfo.BlobRate > 0 {
-			blobRate = rateInfo.BlobRate
+		if rateInfo.BlobRate > unauthRates.PerUserUnauthBlobRate {
+			rates.BlobRate = rateInfo.BlobRate
 		}
 
-		return &PerUserRateInfo{
-			Throughput: throughput,
-			BlobRate:   blobRate,
-		}, nil
+		break
 	}
 
-	return &PerUserRateInfo{
-		Throughput: unauthRates.PerUserUnauthThroughput,
-		BlobRate:   unauthRates.PerUserUnauthBlobRate,
-	}, nil
+	// Check if the address is in the allowlist
+	quorumRates, ok := s.rateConfig.Allowlist[authenticatedAddress]
+	if ok {
+		rateInfo, ok := quorumRates[quorumID]
+		if ok {
+
+			if rateInfo.Throughput > rates.Throughput {
+				rates.Throughput = rateInfo.Throughput
+				keys.ThroughputKey = "address:" + authenticatedAddress
+			}
+
+			if rateInfo.BlobRate > rates.BlobRate {
+				rates.BlobRate = rateInfo.BlobRate
+				keys.BlobRateKey = "address:" + authenticatedAddress
```

Comment:
Can the key just be authenticatedAddress for both throughput and blob rates? They look the same.

Reply:
Yes, I'm going to make this change.